### PR TITLE
Add sq_slots & mp_slots to instance (issue #1197)

### PR DESCRIPTION
**Pull request by sizeoftank**
implement  ` tp_as_sequence ` &  ` tp_as_mapping ` for  ` instance_cls `

### CODE REVIEW

File: src/runtime/classobj.cpp [hunk 1]
```diff
@@ -723,14 +723,115 @@ Box* instanceLen(Box* _inst) {
...
+            return NULL;
+        }
+    } else {
+        getitem_func = _instanceGetattribute(inst, getitem_str, true);
+    }
+
     AUTO_DECREF(getitem_func);
-    return runtimeCall(getitem_func, ArgPassSpec(1), key, NULL, NULL, NULL, NULL);
+    try {
+        return runtimeCall(getitem_func, ArgPassSpec(1), key, NULL, NULL, NULL, NULL);
```
Comment (undingen) [inline src/runtime/classobj.cpp]: I think you can replace the whole try catch block with
```
return runtimeCallInternal<S, NOT_REWRITABLE>(getitem_func, NULL, ArgPassSpec(1), key, NULL, NULL, NULL, NULL);
```

File: src/runtime/classobj.cpp [hunk 2]
```diff
@@ -743,6 +844,39 @@ Box* instanceSetitem(Box* _inst, Box* key, Box* value) {
     return runtimeCall(setitem_func, ArgPassSpec(2), key, value, NULL, NULL, NULL);
 }
 
+static int instance_ass_item(Box* inst, Py_ssize_t k, PyObject* item) noexcept {
+    static BoxedString* delitem_str, *setitem_str;
+    Box* func = NULL, * res = NULL;
+    Box* key = boxInt(k);
+
+    if (item == NULL) {
+        delitem_str = getStaticString("delitem_str");
```
Comment (undingen) [inline src/runtime/classobj.cpp]: `"__delitem__"`

File: src/runtime/classobj.cpp [hunk 3]
```diff
@@ -840,6 +974,78 @@ Box* instanceDelslice(Box* _inst, Box* i, Box* j) {
...
+
+    if (value == NULL) {
+        delslice_str = getStaticString("__delslice__");
+        func = instance_getattro(inst, delslice_str);
+        if (func == NULL) {
+            if (!PyErr_ExceptionMatches(PyExc_AttributeError))
+                return -1;
+            PyErr_Clear();
+
+            delitem_str = getStaticString("delitem_str");
```
Comment (undingen) [inline src/runtime/classobj.cpp]: `"__delitem__"`

**Comment by undingen** [comment]
thanks for the patch! :+1: 
I added a few comments, I'm a little bit worried about the additional code duplication this introduces but it does not make things much worse than the old class support already is :-)
I guess we really just have to replace the whole implementation some point in the future with cpythons but for now this is a improvement.

**Comment by sizeoftank** [comment]
thanks for your response! 👍 
oh, these issues aroused by my careless, I make a change to fix them. :-)
I also think we should refactor our codes to make it better compatibility with cpython, but I not sure I can complete these works correctly now.

File: src/runtime/classobj.cpp [hunk 4]
```diff
@@ -840,6 +966,78 @@ Box* instanceDelslice(Box* _inst, Box* i, Box* j) {
     }
 }
 
+// Analoguous to CPython's, used for sq_slots.
+static int instance_ass_slice(Box* inst, Py_ssize_t i, Py_ssize_t j, PyObject* value) {
+    static BoxedString* delslice_str, *setslice_str, *delitem_str, *setitem_str;
+    Box* func = NULL, * res = NULL;
+    Box* slice = NULL;
+    Box* begin = boxInt(i);
```
Comment (undingen) [inline src/runtime/classobj.cpp]: `begin` and `end` will get leaked at the `return -1` statments.
better use `AUTO_DECREF(begin);`

File: src/runtime/classobj.cpp [hunk 5]
```diff
@@ -840,6 +966,78 @@ Box* instanceDelslice(Box* _inst, Box* i, Box* j) {
...
+            }
+            res = runtimeCallCapi(func, ArgPassSpec(3), begin, end, (Box*)value, NULL, NULL);
+        }
+    }
+
+    Py_DECREF(begin);
+    Py_DECREF(end);
+    Py_DECREF(func);
+
+    if (slice == NULL) {
```
Comment (undingen) [inline src/runtime/classobj.cpp]: this check is strange because if it is NULL it would have already got used....

Comment (sizeoftank) [inline src/runtime/classobj.cpp]: It is a mistake that I copied the cpython implementation to here (it use ` arg = Py_BuildValue() ` every path), I think this check can be removed and change to AUTO_DECREF(slice) when I create it.

**Comment by sizeoftank** [comment]
I stash these into one, other changes need todo?

**Comment by undingen** [comment]
LGTM, thanks for the patch!

**Pull request merged**